[ML] ML stats failures should not stop the usage API working

droberts195 opened the pull request:
It is possible to meddle with internal ML state such that calls to the ML stats APIs return errors. It is justifiable for these single purpose APIs to return errors when the internal state of ML is corrupted. However, it is undesirable for these low level problems to completely prevent the overall usage API from returning, because then callers cannot find out usage information from any part of the system.

This change makes errors in the ML stats APIs non-fatal to the overall response of the usage API. When an ML stats APIs returns an error, the corresponding section of the ML usage information will be blank.

Fixes #91893

droberts195 committed:
[ML] ML stats failures should not stop the usage API working

It is possible to meddle with internal ML state such that calls
to the ML stats APIs return errors. It is justifiable for these
single purpose APIs to return errors when the internal state of
ML is corrupted. However, it is undesirable for these low level
problems to completely prevent the overall usage API from returning,
because then callers cannot find out usage information from any
part of the system.

This change makes errors in the ML stats APIs non-fatal to the
overall response of the usage API. When an ML stats APIs returns
an error, the corresponding section of the ML usage information
will be blank.

Fixes #91893

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

elasticsearchmachine commented:
Hi @droberts195, I've created a changelog YAML for you.

droberts195 committed:
Update docs/changelog/91917.yaml

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearningUsageTransportAction.java:
```diff
@@ -137,45 +141,76 @@ protected void masterOperation(
+            })),
+            e -> {
+                logger.warn("Failed to get job stats to include in ML usage", e);
+                client.execute(GetDatafeedsStatsAction.INSTANCE, datafeedStatsRequest, datafeedStatsListener);
```
This is difficult to read but are you intentionally skipping the call to get job configs if get job stats fails.

droberts195 replied:
Yes, because the `addJobsUsage` method requires both stats and configs, so if either fails then we can't call it.

droberts195 committed:
Skip the YAML test with security enabled

The X-Pack usage endpoint requires the `monitor` privilege, but
we don't want to grant that in our security tests as it would
also grant access to ML endpoints which would interfere with
testing of access rights them.

droberts195 committed:
Merge remote-tracking branch 'droberts195/ml_errors_should_not_block_usage_api' into ml_errors_should_not_block_usage_api

dimitris-athanasiou reviewed (approved):
LGTM
